packages/env/lib/config.js

```diff
@@ -62,200 +77,281 @@ module.exports = {
+				},
+				workDirectoryPath
+			);
+			return result;
+		}, {} );
 
-		// In the future, we should clean this up and integrate it with multi-
-		// environment support instead of hardcoding it to the test port.
-		if ( config.config.WP_TESTS_DOMAIN === undefined ) {
-			config.config.WP_TESTS_DOMAIN = `localhost:${ config.testsPort }`;
-		}
+		return withOverrides( {
+			name: path.basename( configDirectoryPath ),
+			dockerComposeConfigPath: path.resolve(
+				workDirectoryPath,
+				'docker-compose.yml'
+			),
+			configDirectoryPath,
+			workDirectoryPath,
+			...configAll,
+			env,
+		} );
+	},
+};
 
-		if ( config.core !== null && typeof config.core !== 'string' ) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "core" must be null or a string.'
-			);
-		}
+async function getDefaultBaseConfig( configPath ) {
+	const configDirectoryPath = path.dirname( configPath );
+	const type = await detectDirectoryType( configDirectoryPath );
 
-		if (
-			! Array.isArray( config.plugins ) ||
-			config.plugins.some( ( plugin ) => typeof plugin !== 'string' )
-		) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "plugins" must be an array of strings.'
-			);
-		}
+	if ( type === 'core' ) {
+		return { core: '.' };
+	} else if ( type === 'plugin' ) {
+		return { plugins: [ '.' ] };
+	} else if ( type === 'theme' ) {
+		return { themes: [ '.' ] };
+	}
 
-		if (
-			! Array.isArray( config.themes ) ||
-			config.themes.some( ( theme ) => typeof theme !== 'string' )
-		) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "themes" must be an array of strings.'
-			);
-		}
+	throw new ValidationError(
+		`No .wp-env.json file found at '${ configPath }' and could not determine if '${ configDirectoryPath }' is a WordPress installation, a plugin, or a theme.`
+	);
+}
 
-		if ( ! Number.isInteger( config.port ) ) {
+async function getConfigFileData( name, configPath ) {
+	try {
+		return withBackCompat(
+			JSON.parse( await fs.readFile( configPath, 'utf8' ) )
+		);
+	} catch ( error ) {
+		if ( error.code === 'ENOENT' ) {
+			return null;
+		} else if ( error instanceof SyntaxError ) {
 			throw new ValidationError(
-				'Invalid .wp-env.json: "port" must be an integer.'
+				`Invalid ${ name }: ${ error.message }`
 			);
-		}
-
-		if ( ! Number.isInteger( config.testsPort ) ) {
+		} else {
 			throw new ValidationError(
-				'Invalid .wp-env.json: "testsPort" must be an integer.'
+				`Could not read ${ name }: ${ error.message }`
 			);
 		}
+	}
+}
 
-		if ( config.port === config.testsPort ) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "testsPort" and "port" must be different.'
-			);
-		}
+/**
+ * Used to maintain back compatibility with older versions of the .wp-env.json
+ * file. Returns an object in the shape of the currently expected .wp-env.json
+ * version.
+ *
+ * @param {Object} rawConfig config right after being read from a file.
+ */
+function withBackCompat( rawConfig ) {
```

Comment:
with the new format, it doesn't make sense to have both `port` and `testsPort`. So each `WPServiceConfig` now has one `port` property. This function makes sure that existing `.wp-env.json` files continue working.